src/core/Do.Architecture/Architecture/Application.cs

```diff
@@ -20,6 +23,36 @@ internal Application With(ApplicationDescriptor descriptor)
         return this;
     }
 
+    void CheckDuplicates(ApplicationDescriptor descriptor)
+    {
+        try
+        {
+            descriptor.Layers.ToDictionary(
+                keySelector: layer => layer.Id,
+                elementSelector: layer => layer
+            );
+            descriptor.Features.ToDictionary(
+                keySelector: feature => feature.ToString() ?? feature.GetType().Name,
+                elementSelector: feature => feature
+            );
+        }
+        catch (ArgumentException e)
+        {
+            // It takes the name value after "Key:" in the exeception message.
+            string pattern = @"Key:\s*(.*?)(?:\s|$)";
+
+            Match match = Regex.Match(e.Message, pattern);
+
+            string key = string.Empty;
+            if (match.Success)
+            {
+                key = match.Groups[1].Value;
+            }
+
+            throw new InvalidOperationException($"Cannot add `{key}`, it was already added.");
```

Comment:
because of the bad smell above, this exception handling gets complicated, it will simplify when you find a better way there